### PR TITLE
Test and fix for issue #354

### PR DESCRIPTION
**Pull request by mlhetland**
Incorporates @timholy's fix and adds my test.

Note: The test does not properly exercize the added `yield()` statement.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -2162,6 +2162,40 @@ end
...
+    modname = "A354"
+    srcfile = joinpath(testdir, modname * ".jl")
+
+    function setvalue(x)
+        open(srcfile, "w") do io
+            print(io, "module $modname test() = $x end")
+        end
+    end
+
+    setvalue(1)
```
Comment (timholy) [inline test/runtests.jl]: I think you want `sleep(mtimedelay)` before & after the `@eval using` for reasons we discussed in #354. I know these tests don't run on OSX by default but we might as well get them as close to passing as possible.

**Comment by mlhetland** [comment]
Hm. It seems the checks are failing…? Both warnings about running out of inotify capacity and an undefined `rseed`…?

File: test/runtests.jl (hunk 1 above)
Comment (mlhetland) [reply test/runtests.jl]: Right. This one actually *does* run on macOS at the moment, as it seemed to work just fine on my machine – but I can add the same conditional around it as the other `entr` stuff, if you'd like. As for the sleep – I didn't think it applied at that point in the code? (I'm just importing the module for the first time?) But no problem, I'll add that.

**Comment by mlhetland** [comment]
(Oops, it seems my use of `@eval` in the commit msg. tagged a user. Sorry!)

File: test/runtests.jl (hunk 1 above)
Comment (timholy) [reply test/runtests.jl]: Maybe you were right, didn't notice the sleep after the `@async` loop. Still can't hurt.

**Comment by mlhetland** [comment]
Well, I can't really make any sense of the check failures, here.

**Comment by timholy** [comment]
Does it reproduce locally if you do `]test Revise`?

**Comment by timholy** [comment]
The `inotify` text is expected, part of tests that only run on Linux & Travis. See old passing builds in the Travis history.

**Comment by mlhetland** [comment]
It works locally – tested that before creating the PR. Also, I don't quite see the link between the errors and the new code/test?

**Comment by timholy** [comment]
It fails for me locally. Are you sure you tested it with `]test Revise`? (The `]` meaning pkg-mode.)

**Comment by mlhetland** [comment]
Yes, I understood. I was sure it worked at some point, but I can see it doesn't, now. (It did work when run from the command-line without `]`.)

**Comment by mlhetland** [comment]
The new test succeeds, though, and the failure seems unrelated to `entr`.

**Comment by mlhetland** [comment]
Oh, well, I'll do some local debugging :-)

**Comment by mlhetland** [comment]
Well, at least it seems to be specifically the use of `entr` that fails – not the call to `A354.test()` nor any other part of the test (such as the `@async` stuff). And it fails with both the new and the old version of `entr`. However, if I run `julia test/runtests.jl`, it works just fine. And all I get from `Pkg` is:
```
ERROR: Package Revise errored during testing
```
Is there no way of getting the full stack trace?

**Comment by timholy** [comment]
Definitely a weird error. When I run at the REPL, I see it starting the cleanup phase and then Julia just quits outright, dumping me back to the Linux prompt. Might be a Julia bug. I suspect it's a task problem.

**Comment by timholy** [comment]
Read about the complexities at https://github.com/JuliaLang/julia/issues/6283 and then change the test so it doesn't try to stop the task.

**Comment by mlhetland** [comment]
Ah, right. I actually came across that in my quest for how to stop the task without using exceptions. And I didn't look at it thoroughly enough :-}

The reason I didn't want to use exceptions was that I seemed to end up with a compound exception that I had to unpack (similarly to in the other test), though I thought that throwing an `InterruptException` should work in just terminating `entr`.

And now, somehow, it does… So unless there's some reason not to, I'll just go with that.

**Comment by mlhetland** [comment]
Whoah! Seemed to more or less for the early checks, but then triggered a segfault in Julia nightly! Any ideas?

**Comment by timholy** [comment]
Yay! The segfault may be #344

But I get the same error that shows up in one of the appveyor tests, https://ci.appveyor.com/project/timholy/revise-jl/builds/27252479/job/d20rpoxf78awybn4. Looking at the code, I am a bit surprised it works; I would have thought your implementation would suffer from scope problems. Maybe make `result` a `Ref`? I.e., `result = Ref(0)` and then access it with `result[]`.

**Comment by mlhetland** [comment]
Hm. Could there still be scope issues, even when using a `Ref`? I.e., could there be some copying going on, or something? Maybe I'll try to have a global result, instead…

**Comment by mlhetland** [comment]
OK, trying to move the `entr` and result assignment out of the `@async` (beyond what's in `entr` to begin with; I guess nesting that was an issue to begin with), moving the file modification to a separate task instead. Hopefully a bit cleaner.

And now I remember why the `InterruptException` didn't work with `entr` before … it's because I didn't have it in a separate task. So now, once again, that doesn't work. Given that we should only get the correct result assignment if nothing goes wrong, I'm just catching my own `error` with an indiscriminate `catch` here. Could do more checking and `rethrow`ing, if you think that's necessary.

**Comment by mlhetland** [comment]
Hm. And now it stalls (sometimes) – I guess the `entr`wasn't triggered, then. Maybe I'll try another `sleep`…? :-P

**Comment by mlhetland** [comment]
Hm. The stall was on macOS, where you've set `mtimedelay` to 2.1, so the wait before `setvalue` should be plenty of time to get `entr` running? I dunno. Adding a bit more poking at the file, just to see if it helps.

**Comment by mlhetland** [comment]
Welp, at least now the only issues are the Julia nightly segfaults, as far as I can see. So I guess things are perhaps good to go? (Unless you want to turn off this new test on macOS, like the other `entr` test; it works, for now, but…)

**Comment by timholy** [comment]
This indeed looks fine. Thanks for your patience! While the final result is fairly simple, it's pretty amazing how many factors come into place when you combine watching file systems & recompilation.

**Comment by mlhetland** [comment]
That's fine! Yes, it's indeed a bit tricky – and here we had issues with `@async` mixed up with it all, I guess :-) And the resulting test is sort of a test we might have wanted for `entr` to begin with, I guess.

**Pull request merged**